### PR TITLE
[Onprem] Script for Automatic setup/launching for Onprem servers

### PR DESCRIPTION
**Pull request by michaelzhiluo**
Should be tested after #1114 and #1116 are merged. This PR releases a script that automatically launches a cloud VM (acting as an onprem server) and sets up a regular user called `test` on the cloud server. It will also run `sky admin deploy` and setup the local cluster config as needed.

## Example

`python examples/local/launch_cloud_onprem.py -- local-cluster-name hello`
`sky launch -c hello -- 'echo hi'`

### CODE REVIEW

File: examples/local/launch_cloud_onprem.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import argparse
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:1]: Add a comment to top explaining what the script is and how to use it?

File: examples/local/launch_cloud_onprem.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+from click import testing as cli_testing
+
+from sky import cli
+from sky import global_user_state
+from sky.backends import onprem_utils
+from sky.utils import common_utils
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--local-cluster-name',
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:17]: If this is a required arg, make it a positional arg? or add a short hand `-n`

File: examples/local/launch_cloud_onprem.py (hunk 1 above)
Comment (romilbhardwaj) [reply examples/local/launch_cloud_onprem.py:1]: Something like:
```
"""Launches an onprem cluster for testing purposes. 

<more details on how it works/ what it does. Also say something about how to clean up the created the cluster>

Usage:
  <usage example>
"""
```

File: examples/local/launch_cloud_onprem.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    subprocess.run(f'sky admin deploy {file_path}', shell=True, check=True)
+
+# Fill out the local config file in ~/.sky/local/...
+local_config = onprem_utils.get_local_cluster_config_or_error(
+    local_cluster_name)
+local_config['auth']['ssh_user'] = 'test'
+local_config['auth']['ssh_private_key'] = private_key
+local_config_path = onprem_utils.SKY_USER_LOCAL_CONFIG_PATH.format(
+    local_cluster_name)
+common_utils.dump_yaml(os.path.expanduser(local_config_path), local_config)
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:81]: Maybe add a nice print at the end saying something along the lines of `Onprem cluster {name} is now ready for use! You can launch jobs with <cmd>. After you are down, shut down by running <cmds>`

File: examples/local/launch_cloud_onprem.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+              sudo -H su test -c 'cat /user-key >> ~/.ssh/authorized_keys'
+            """)
+}
+
+cli_runner = cli_testing.CliRunner()
+onprem_name = f'onprem-cluster-{uuid.uuid4().hex[:6]}'
+with tempfile.NamedTemporaryFile(suffix='.yaml', mode='w') as f:
+    yaml.dump(sky_yaml_config, f)
+    file_path = f.name
+    cli_runner.invoke(cli.launch, ['-c', onprem_name, file_path])
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:52]: This failed for me :(
```
(base) romilb@romilbx1yoga:/mnt/d/Romil/Berkeley/Research/sky-experiments/examples/local$ python launch_cloud_onprem.py --local-cluster-name myclus
I 08-26 20:04:35 optimizer.py:605] == Optimizer ==
I 08-26 20:04:35 optimizer.py:628] Estimated cost: $0.4 / hour
I 08-26 20:04:35 optimizer.py:628]
I 08-26 20:04:35 optimizer.py:685] Considered resources (1 node):
I 08-26 20:04:35 optimizer.py:713] ------------------------------------------------------------------
I 08-26 20:04:35 optimizer.py:713]  CLOUD   INSTANCE      vCPUs   ACCELERATORS   COST ($)   CHOSEN
I 08-26 20:04:35 optimizer.py:713] ------------------------------------------------------------------
I 08-26 20:04:35 optimizer.py:713]  AWS     m6i.2xlarge   8       -              0.38          ✔
I 08-26 20:04:35 optimizer.py:713] ------------------------------------------------------------------
I 08-26 20:04:35 optimizer.py:713]
I 08-26 20:04:35 cloud_vm_ray_backend.py:2604] Creating a new cluster: "onprem-cluster-7ebe5a" [1x AWS(m6i.2xlarge)].
I 08-26 20:04:35 cloud_vm_ray_backend.py:2604] Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
I 08-26 20:04:45 cloud_vm_ray_backend.py:892] To view detailed progress: tail -n100 -f /home/romilb/sky_logs/sky-2022-08-26-20-04-34-484524/provision.log
I 08-26 20:04:47 cloud_vm_ray_backend.py:1096] Launching on AWS us-east-1 (us-east-1a,us-east-1b,us-east-1c,us-east-1d,us-east-1e,us-east-1f)
I 08-26 20:06:15 log_utils.py:45] Head node is up.
I 08-26 20:07:15 cloud_vm_ray_backend.py:984] Successfully provisioned or found existing VM.
I 08-26 20:07:16 cloud_vm_ray_backend.py:2648] Processing file mounts.
I 08-26 20:07:23 cloud_vm_ray_backend.py:2677] To view detailed progress: tail -n100 -f ~/sky_logs/sky-2022-08-26-20-04-34-484524/file_mounts.log
I 08-26 20:07:23 backend_utils.py:996] Syncing (to 1 node): ~/.ssh/sky-key.pub -> ~/.sky/file_mounts/user-key
I 08-26 20:07:24 cloud_vm_ray_backend.py:1910] Running setup on 1 node.
Adding user `test' ...
Adding new group `test' (1001) ...
Adding new user `test' (1001) with group `test' ...
Creating home directory `/home/test' ...
Copying files from `/etc/skel' ...
I 08-26 20:07:26 cloud_vm_ray_backend.py:1913] Setup completed.
I 08-26 20:07:26 cloud_vm_ray_backend.py:2072] Run commands not specified or empty.
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100]
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100] Cluster name: onprem-cluster-7ebe5a
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100] To log into the head VM: ssh onprem-cluster-7ebe5a
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100] To submit a job:         sky exec onprem-cluster-7ebe5a yaml_file
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100] To stop the cluster:     sky stop onprem-cluster-7ebe5a
I 08-26 20:07:26 cloud_vm_ray_backend.py:2100] To teardown the cluster: sky down onprem-cluster-7ebe5a
NAME                   LAUNCHED     RESOURCES            STATUS  AUTOSTOP  COMMAND                       
onprem-cluster-7ebe5a  16 secs ago  1x AWS(m6i.2xlarge)  UP      -         launch_cloud_onprem.py --...

Local clusters:
NAME          USER  HEAD_IP        RESOURCES    COMMAND
onprem2       test  54.186.202.57  [(no GPUs)]  sky launch -c onprem2 min...
azure-onprem  -     -              -            -
onprem-test   -     -              -            -
[1/4] Checking on-premise environment
E 08-26 20:07:39 subprocess_utils.py:69] 
sky.exceptions.CommandError: Command sudo ray --version failed with return code 1.
Ray is not installed on 54.89.85.189.
Traceback (most recent call last):
  File "launch_cloud_onprem.py", line 72, in <module>
    subprocess.run(f'sky admin deploy {file_path}', shell=True, check=True)
  File "/mnt/d/wsl/anaconda3/lib/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'sky admin deploy /tmp/tmp7_oqop5b.yaml' returned non-zero exit status 1.
```

**Review by romilbhardwaj — COMMENTED** [review]
Thanks for adding! This would make it really easy for us to setup our own onprem debugging clusters.

File: examples/local/launch_cloud_onprem.py (hunk 1 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:1]: Done, thanks!

File: examples/local/launch_cloud_onprem.py (hunk 4 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:52]: Ah this PR should be tested after #1114 and #1116 are merged!

File: examples/local/launch_cloud_onprem.py (hunk 2 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:17]: Done, thanks!

File: examples/local/launch_cloud_onprem.py (hunk 3 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:81]: Done, thanks!

**Review by romilbhardwaj — COMMENTED** [review]
Thanks for addressing previous comments! I tried it out and still getting the error mentioned [here](https://github.com/skypilot-org/skypilot/pull/1115#discussion_r956529686). As you mentioned, this is because it has a dependency on #1114 and #1116. We can rebase/merge this PR on #1116 (after it is merged/rebased on #1114) so we can start testing this

File: examples/local/launch_cloud_onprem.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+from sky import global_user_state
+from sky.backends import onprem_utils
+from sky.utils import common_utils
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-n',
+                    '--local-cluster-name',
+                    type=str,
+                    required=True,
+                    help='Name of the local cluster.')
```
[the remark this answers is not in the document]
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py]: Done, thanks!

**Comment by michaelzhiluo** [comment]
@romilbhardwaj Just tested the script. It works as of now.

File: examples/local/launch_cloud_onprem.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+"""Automatically launches and sets up a Sky onprem cluster on AWS.
+
+The script allocates a cloud cluster and setups up the cloud cluster to run an
+'onprem' cluster on the cloud. It performs the following steps:
+    1) Launches a cluster on the cloud with `sky launch -c on-prem-cluster-[ID]`.
+     The Sky yaml sets up another user named `test`.
+    2) Setups Sky dependencies for the admin user via `sky admin deploy ..`.
+    3) Creates the local cluster config for the regular user, stored in `~/.sky/local/...`
+To clean up the local cluster, run `sky down [LOCAL_CLUSTER]` and remove the corresponding
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:9]: seems like `sky down [cluster_name]` is not sufficient - it just downs the onprem cluster, but not the actual VM running the cluster. For instance, I ran `python launch_cloud_onprem.py myclus`, but even after `sky down myclus` I have the host VM running:

```
(base) romilb@romilbx1yoga:/mnt/d/Romil/Berkeley/Research/sky-experiments/examples/onprem_test$ sky status
NAME                   LAUNCHED    RESOURCES            STATUS  AUTOSTOP  COMMAND                       
onprem-cluster-91644f  13 hrs ago  1x AWS(m6i.2xlarge)  UP      -         launch_cloud_onprem.py my...  
```

I believe we should fetch the autogenerated name add `sky down <name, eg. onprem-cluster-91644f>` to L103.

**Review by romilbhardwaj — COMMENTED** [review]
Worked nicely, thanks @michaelzhiluo! Just need to add some more hints to the user on how to properly shut down the onprem cluster, after that we should be good to go!

File: examples/local/launch_cloud_onprem.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+local cluster config file.
+
+Usage:
+    # Creates a Sky on-premise cluster named `my-local-cluster`
+    python examples/local/launch_cloud_onprem.py -n my-local-cluster
```
[the remark this answers is not in the document]
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:14]: Done, thanks!

File: examples/local/launch_cloud_onprem.py (hunk 6 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py:9]: Donwe, thanks!

**Comment by michaelzhiluo** [comment]
@romilbhardwaj Done, think its ready to be merge 👍

File: examples/local/launch_cloud_onprem.py [hunk 8]
```diff
@@ -100,5 +102,5 @@
 print((
     f'Sky onprem cluster {local_cluster_name} is now ready for use! '
     f'You can launch jobs with `sky launch -c {local_cluster_name}  -- [CMD]. '
```
Comment (romilbhardwaj) [inline examples/local/launch_cloud_onprem.py:104]: nit: there's a double space and missing `:

```
`sky launch -c {local_cluster_name} -- [CMD]`.
```

**Review by romilbhardwaj — APPROVED** [review]
lgtm, thanks for adding!

File: examples/local/launch_cloud_onprem.py (hunk 8 above)
Comment (michaelzhiluo) [reply examples/local/launch_cloud_onprem.py]: Done, thanks!

**Pull request merged**